### ISSUE TITLE
Haneefa : separation-of-concerns, 3 weeks

### ISSUE THREAD

**Issue by Haneefa-Shaik** [lab-brussels-1/home#375]
## Learning Objectives

<details>
<summary>Priorities: 🥚, 🐣, 🐥, 🐔 (click to learn more)</summary>
<br>

There is a lot to learn in this repository. If you can't master all the material
at once, that's expected! Anything you don't master now will always be waiting
for you to review when you need it. These 4 emoji's will help you prioritize
your study time and to measure your progress:

- 🥚: Understanding this material is required, it covers the base skills you'll
  need to move on. You do not need to finish all of them but should feel
  comfortable that you could with enough time.
- 🐣: You have started all of these exercises and feel you could complete them
  all if you just had more time. It may not be easy for you but with effort you
  can make it through.
- 🐥: You have studied the examples and started some exercises if you had time.
  You should have a big-picture understanding of these concepts/skills, but may
  not be confident completing the exercises.
- 🐔: These concepts or skills are not necessary but are related to this module.
  If you are finished with 🥚, 🐣 and 🐥 you can use the 🐔 exercises to push
  yourself without getting distracted from the module's main objectives.

---

</details>

- [ ] 🥚 **event-driven programming**: identify the concept in a JS program via listeners & handlers
- [ ]🥚 **entry points**: describe what an entry point is, there are 2 kinds in the programs for this module - initialization & interaction. identify them in a program
- 🥚 **function roles**: describe what function roles are and why they're important. they can identify a function's role given checklists for each role covered in this module:
  - [ ] **_listeners_**: functions that attach event listeners to the DOM
  - [ ] 🥚 **_handlers_**: entry point for user interactions
  - [ ] 🥚 **_utils_** _(utilities)_: pure functions to help do things with data
  - [ ] 🐣 **_components_**: render data into DOM elements to display for the user
  - [ ] 🐥 **_custom events_**: create custom events events in your components, passing useful data between components and handlers
- [ ] 🥚 **DOM access**: You read and write values from the DOM in an event handler
- [ ] 🥚 **es5 vs. es6**: You can demonstrate the change in developer-experience pre and post es6 by stepping through in the debugger and explaining differences in scoping \(global vs. script, block vs. local, modules\) between two programs with identical user experience but different implementations.
- [ ] 🥚 **Scope Hierarchy**: You is comfortable navigating different scopes in the browser's debugger to understand an existing application \(script, module, closure, local, block\)
- [ ] 🥚 **Code Splitting**: You can use ES Modules to split your code into multiple files & folders according to function role, data, listeners and initialization. They can use generated dependency diagrams and documentation to understand and navigate this folder structure.
- [ ] 🥚 **Dependency Graphs**: You can use a project's dependency graph to understand how it is organized and to navigate the source code.
- [ ] 🥚 **Development Strategies**: You can write development strategies that have all of the program's state defined at the beginning, and separate each user story into _interface_ and _interaction_ tasks.
- [ ] 🐣 **Naming Functions**: You can come up with clear and helpful names for the functions in your program. A good function name will take into account the function's role and the program's domain, like in the `/naming-variables` exercises from Debugging.
- [ ] 🐣 **DOM manipulation**: You can manipulate the DOM when implementing level-appropriate user interactions
- [ ] 🐣 **Isolating Components**: You can use a `test.html` file to render your components with different inputs
- [ ] 🐣 **Forms**: You can do basic handling of form data via `event.target.form`
- [ ] 🐣 **Handling events**: You can use the `event` argument to process user interactions, including bubbled events using `event.target`
- [ ] 🐣 **Passing Component Unit Tests**: You can write vanilla DOM component functions to pass provided unit tests
- [ ] 🐣 **refactoring**: refactor a single-script tutorial-style web page into multiple files using imports and exports
- [ ] 🐥 **reverse-engineering**: You can incrementally reverse-engineer a level-appropriate user interaction following these steps:
  - _init_
  - Listeners
  - Handlers
  - (possibly): Utils, Components, Custom Events
- [ ] 🐔 **From Spec**: given user stories, You can develop a site from scratch using a template repository.
- [ ] 🐔 **Writing Component Unit Tests**: You can write unit tests to validate your component functions using BDD syntax

**Comment by Haneefa-Shaik** [comment]
Week 1

## I Need Help With:

This week, to be honest, the topics are very complex to understand. Did a lot of googling, watched the recorded class video again. https://medium.com/geekculture/event-handlers-vs-event-listeners-in-javascript-b4086b8040b0  In this link, i understood about listeners and handlers. I'm getting more difficulty in understanding about Components.


## What went well?
Done few exercises and studying more about DOM manipulation .

## What went less well?
Still im trying to understand deeper about DOM manipulation. Its really take a lot of time to understand. Trying watch some traversy videos.


## Sunday Prep Work
Still in the completion of exercises.

**Comment by danielhalasz** [comment]
> I'm getting more difficulty in understanding about Components.

that's okay, don't worry about them, we will learn more about them in the module: Component Based Design / React.

**Comment by danielhalasz** [comment]
> Done few exercises and studying more about DOM manipulation .

understanding DOM manipulation is one part of the module..but equally important is to understand separation of concerns

**Comment by Haneefa-Shaik** [comment]
Week 2

## I Need Help With:

The first time, took a little bit of time to check in.  separating files for the existing code is now a little bit habituated by doing the exercises. But for the group project, writing my own code is a bit complicated for me and I'm trying my best.



## What went well?
- practicing and studying last week's DOM manipulation was really helpful this week.
- understood the concept of Separation of concerns.
- Now I understood the dependency graph a little bit clearer when compared to last week.

## What went less well?
- I should admit that I'm less active in this module. Initially, I was a little confused and overwhelmed by these concepts. 
- Even though I practiced the exercises and saw tutorials, still i am not fully learned and I know that there is a lot to go.
- Hopefully, this project will gain confidence in me.


## Sunday Prep Work
on the project work.

**Comment by danielhalasz** [comment]
> https://medium.com/geekculture/event-handlers-vs-event-listeners-in-javascript-b4086b8040b0

just got to read through this..and I would say it would be better to forget about it 🙂 it is confusing and not using them how it is suggested..handlers and listeners are not interchangeable..as that author suggests..but work hand in hand. the listener triggering the handler function..and of course multiple event can trigger the same handler function..for example..a mouse click on the submit button or pressing enter on the keyboard.

**Comment by Haneefa-Shaik** [comment]
Thanks for the clear explanation @danielhalasz

**Comment by Haneefa-Shaik** [comment]
Week 3

## I Need Help With:

This week, doing exercises helped in a good way. Doing the project helped in understand a bit more about the separation of files.


## What went well?
- project work went well
- spending time doing exercises more this week
- practicing DOM

## What went less well?
- Still separation of files on my own is difficult for me. referring the previous exercise and was able to do and understand the concept. writing by own still stucks.


## Sunday Prep Work
on the project work.

**Comment by Haneefa-Shaik** [comment]
## Me

### What was not clear, where did you get stuck?

First two weeks I was overwhelmed by these concepts and was less active and low in confidence. the last week, I have put a little bit of effort and studying. Now at least I am little bit confidence than before

### What was clear, what did you master?

As of now, I'm familiar with Dom manipulation and little bit of code splitting.


### Where can you help others moving forwards?

I am always ready to help others with the concepts which I am capable of making them understand.

 ---

## The Course

### What can there be more of?

May be reviewing the work done by the students

### What can there be less of?

As of now nothing

### What material were most helpful (from HYF or elsewhere)?

As usual, the study lenses are awesome. only the thing is to make time and study this good material.

### Any suggestions for future classes?
 everything is good now including extra study hour

**Comment by danielhalasz** [comment]
> * Still separation of files on my own is difficult for me. referring the previous exercise and was able to do and understand the concept. writing by own still stucks.

we can have a short zoom call sometime to talk about any remaining confusing parts

**Comment by danielhalasz** [comment]
> May be reviewing the work done by the students

yes, that is very important, I agree. once you do a PR or there is work ready to be reviewed by a coach, write them a message on Slack, or tag them on GitHub, so they know about it
